### PR TITLE
feat(common): BREAKING CHANGE add support for PVC static provisioning and dataSource

### PR DESCRIPTION
**Pull request by Ornias1993**
**Description**
This PR aims to add support for static provisioning of PVC-PV volume pairs.

THis ensures we're in-line with what iX-Common backend for SCALE supports and expands on that by allowing users on non-SCALE systems to create custom PVC-PV bindings

It also adds initial support for dataSource, for either backup/snapshot or PVC cloning purposses.
In addition it ensure that a randomised hash is used when dataSource is set, so we can force a recrease when it's changed.

This also merged `.Values.volumeClaimTemplates` with `.Values.persistence` by setting `.type` to `vct` on any persistence reference.
This means users with statefull sets, need to move **all** their `.Values.volumeClaimTemplates` references into `.Values.persistence` like this:

old:
```
volumeClaimTemplates:
  data:
    enabled: true
    mountPath: "/var/lib/ldap/"
```

New:
```
persistence:
  data:
    enabled: true
    type: "vct"
    mountPath: "/var/lib/ldap/"
```

On top of this, this PR breaks ALL VCT's by improving the naming scheme.
It's important to NOT install this update on statefull sets lightly, as this might destroy your data.

Fixes: #558 
Fixes: #562

**⚙️ Type of change**

- [x] ⚙️ Feature/App addition
- [ ] 🪛 Bugfix
- [x] ⚠️ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 🔃 Refactor of current code

**🧪 How Has This Been Tested?**

**📃 Notes:**
Technically breaking, as this changes the CNPG storage structure in values.yaml.
The new structure for CNPG storage is the same as .Values.Persistence (type is always PVC)

Also Breaking because of VolumeClaimTemplates being merged with Persistence

**✔️ Checklist:**

- [x] ⚖️ My code follows the style guidelines of this project
- [x] 👀 I have performed a self-review of my own code
- [x] #️⃣ I have commented my code, particularly in hard-to-understand areas
- [x] 📄 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] 🧪 I have added tests to this description that prove my fix is effective or that my feature works
- [x] ⬆️ I increased versions for any altered app according to semantic versioning

**➕ App addition**

If this PR is an app addition please make sure you have done the following.

- [x] 🪞 I have opened a PR on [truecharts/containers](https://github.com/truecharts/containers) adding the container to TrueCharts mirror repo.
- [x] 🖼️ I have added an icon in the Chart's root directory called `icon.png`

---

_Please don't blindly check all the boxes. Read them and only check those that apply.
Those checkboxes are there for the reviewer to see what is this all about and
the status of this PR with a quick glance._

### CODE REVIEW

**Comment by Ornias1993** [comment]
Mostly required: Review and test fixes.

File: library/common/templates/lib/container/_volumeMounts.tpl [hunk 1]
```diff
@@ -12,9 +12,6 @@ objectData: The object data to be used to render the container.
 
   {{- $codeServerIgnoredTypes := (list "configmap" "secret") -}}
   {{- $keys := (list "persistence") -}}
-  {{- if eq $objectData.podType "StatefulSet" -}}
-    {{- $keys = mustAppend $keys "volumeClaimTemplates" -}}
-  {{- end -}}
 
   {{- range $key := $keys -}}
```
Comment (stavros-k) [inline library/common/templates/lib/container/_volumeMounts.tpl:16]: We should remove the range $key and skip vct type if its not STS

Comment (Ornias1993) [reply library/common/templates/lib/container/_volumeMounts.tpl]: Done the skip part...

Comment (Ornias1993) [reply library/common/templates/lib/container/_volumeMounts.tpl]: Done both

File: library/common/templates/lib/container/_volumeMounts.tpl [hunk 2]
```diff
@@ -10,21 +10,15 @@ objectData: The object data to be used to render the container.
...
-        {{/* Dont try to mount configmap/sercet to codeserver */}}
-        {{- if not (and (eq $objectData.shortName "codeserver") (mustHas $persistenceValues.type $codeServerIgnoredTypes)) -}}
-          {{- $volMount := (fromJson (include "tc.v1.common.lib.container.volumeMount.isSelected" (dict "persistenceName" $persistenceName "persistenceValues" $persistenceValues "objectData" $objectData "key" $key))) -}}
-          {{- if $volMount -}}
-            {{- $volMounts = mustAppend $volMounts $volMount -}}
-          {{- end -}}
+  {{- $codeServerIgnoredTypes := (list "configmap" "secret" "vct") -}}
+
+  {{- range $persistenceName, $persistenceValues := $rootCtx.Values.persistence -}}
+    {{- if and $persistenceValues.enabled ( and ( eq $persistenceValues.type "vct") ( ne $objectData.podType "StatefulSet" ) )  -}}
```
Comment (stavros-k) [inline library/common/templates/lib/container/_volumeMounts.tpl:16]:
```suggestion
    {{- if and $persistenceValues.enabled (not (and (eq $persistenceValues.type "vct") (ne $objectData.podType "StatefulSet"))) -}}
```

image

File: library/common/templates/spawner/_pvc.tpl [hunk 3]
```diff
@@ -19,23 +18,64 @@
...
+      {{- if and (mustHas $objectData.type $types) (not $objectData.existingClaim) -}}
 
-        {{- $objectName := (printf "%s-%s" $fullname $name) -}}
-        {{/* Perform validations */}}
-        {{- include "tc.v1.common.lib.chart.names.validation" (dict "name" $objectName) -}}
-
-        {{/* Set the name of the secret */}}
-        {{- $_ := set $objectData "name" $objectName -}}
+        {{/* Set the name of the PVC */}}
+        {{- $_ := set $objectData "name" ( include "tc.v1.common.lib.storage.pvc.name" (dict "rootCtx" $ "objectName" $objectName "objectData" $objectData)) -}}
```
Comment (stavros-k) [inline library/common/templates/spawner/_pvc.tpl:26]:
```suggestion
        {{- $_ := set $objectData "name" (include "tc.v1.common.lib.storage.pvc.name" (dict "rootCtx" $ "objectName" $name "objectData" $objectData)) -}}
```

File: library/common/templates/lib/storage/pvc/_spec.tpl [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+*/}}
+{{- define "tc.v1.common.lib.storage.pvc.spec" -}}
+{{- $rootCtx := .rootCtx -}}
+{{- $objectData := .objectData -}}
+
+accessModes:
+  {{- include "tc.v1.common.lib.pvc.accessModes" (dict "rootCtx" $rootCtx "objectData" $objectData "caller" "PVC") | trim | nindent 4 }}
+resources:
+  requests:
+    storage: {{ $objectData.size | default $.Values.fallbackDefaults.pvcSize }}
```
Comment (stavros-k) [inline library/common/templates/lib/storage/pvc/_spec.tpl:15]:
```suggestion
    storage: {{ $objectData.size | default $rootCtx.Values.fallbackDefaults.pvcSize }}
```

**Pull request merged**